launcher/firstLaunch/firstlaunch_moc.cpp

```diff
@@ -338,248 +402,317 @@ void FirstLaunchView::extractGogData()
-	if(!errorText.isEmpty())
-		hashError = Innoextract::getHashError(tmpFileExe, tmpFileBin, filePathExe, filePathBin);
+        if(tgt.compare("Data", Qt::CaseInsensitive)!=0 && tgt.compare("Maps", Qt::CaseInsensitive)!=0 && tgt.compare("Mp3",  Qt::CaseInsensitive)!=0)
+            continue;
 
-	QStringList dirData = tempDir.entryList({"data"}, QDir::Filter::Dirs);
-	if(!errorText.isEmpty() || dirData.empty() || QDir(tempDir.filePath(dirData.front())).entryList({"*.lod"}, QDir::Filter::Files).empty())
-	{
-		if(!errorText.isEmpty())
-		{
-			logGlobal->error("Gog installer extraction failure! Reason: %s", errorText.toStdString());
-			QMessageBox::critical(this, tr("Extracting error!"), errorText, QMessageBox::Ok, QMessageBox::Ok);
-			if(!hashError.isEmpty())
-			{
-				logGlobal->error("Hash error: %s", hashError.toStdString());
-				QMessageBox::critical(this, tr("Hash error!"), hashError, QMessageBox::Ok, QMessageBox::Ok);
-			}
-		}
-		else
-			QMessageBox::critical(this, tr("No Heroes III data!"), tr("Selected files do not contain Heroes III data!"), QMessageBox::Ok, QMessageBox::Ok);
-		tempDir.removeRecursively();
-		return;
-	}
-
-	logGlobal->info("Copying provided game files...");
-	copyHeroesData(tempDir.path(), true);
-
-	tempDir.removeRecursively();
-#endif
-}
-
-void FirstLaunchView::copyHeroesData(const QString & path, bool move)
-{
-	QDir sourceRoot{path};
-
-#ifdef VCMI_IOS
-	// TODO: Qt 6.5 can select directories https://codereview.qt-project.org/c/qt/qtbase/+/446449
-	SelectDirectory iosDirectorySelector;
-	if(path.isEmpty())
-		sourceRoot.setPath(iosDirectorySelector.getExistingDirectory());
-#else
-	if(path.isEmpty())
-		sourceRoot.setPath(QFileDialog::getExistingDirectory(this, {}, {}, QFileDialog::ShowDirsOnly | QFileDialog::DontResolveSymlinks));
-#endif
-
-	if(!sourceRoot.exists())
-		return;
-
-	if (sourceRoot.dirName().compare("data", Qt::CaseInsensitive) == 0)
-	{
-		// We got Data folder. Possibly user selected "Data" folder of Heroes III install. Check whether valid data might exist 1 level above
-
-		QStringList dirData = sourceRoot.entryList({"data"}, QDir::Filter::Dirs);
-		if (dirData.empty())
+        if(!created.contains(tgt))
 		{
-			// This is "Data" folder without any "Data" folders inside. Try to check for data 1 level above
-			sourceRoot.cdUp();
-		}
-	}
+            QDir{}.mkpath(targetRoot.filePath(tgt));
+            created.insert(tgt);
+        }
 
-	QStringList dirData = sourceRoot.entryList({"data"}, QDir::Filter::Dirs);
-	QStringList dirMaps = sourceRoot.entryList({"maps"}, QDir::Filter::Dirs);
-	QStringList dirMp3 = sourceRoot.entryList({"mp3"}, QDir::Filter::Dirs);
+        const QDir dstDir = targetRoot.filePath(tgt);
+        plan.push_back({ src, dstDir.filePath(file) });
+    }
 
-	const auto noDataMessage = tr("Failed to detect valid Heroes III data in chosen directory.\nPlease select the directory with installed Heroes III data.");
-	if(dirData.empty())
+    if(plan.isEmpty())
```

Comment:
you have already detected that H3 installation is a valid SoD, this will never be true